File: scan-livros.js
const fs = require('fs');
const path = require('path');
const { parseStringPromise } = require('xml2js');

// Constantes de configuração
const LIVROS_PATH = path.join(__dirname, 'Livros');
const OUTPUT_FILE = path.join(__dirname, 'catalogo.json');
const METADATA_FILENAME = 'metadata.opf';
const COVER_FILENAME = 'cover.jpg';
const CALIBRE_TRASH_FOLDER = '.caltrash';
const SUPPORTED_FORMATS = ['.pdf', '.epub', '.mobi'];

/**
 * Busca recursivamente todos os arquivos metadata.opf na pasta especificada
 * @param {string} dir - Diretório raiz para busca
 * @param {Array<string>} arquivosEncontrados - Array acumulador de arquivos encontrados
 * @returns {Array<string>} Lista de caminhos completos dos arquivos metadata.opf
 */
function buscarArquivosOPF(dir, arquivosEncontrados = []) {
    const arquivos = fs.readdirSync(dir);
    
    for (const arquivo of arquivos) {
        if (arquivo === CALIBRE_TRASH_FOLDER) continue;
        
        const caminhoCompleto = path.join(dir, arquivo);
        const isDirectory = fs.statSync(caminhoCompleto).isDirectory();
        
        if (isDirectory) {
            buscarArquivosOPF(caminhoCompleto, arquivosEncontrados);
        } else if (arquivo === METADATA_FILENAME) {
            arquivosEncontrados.push(caminhoCompleto);
        }
    }
    
    return arquivosEncontrados;
}

/**
 * Extrai informações do livro a partir do arquivo metadata.opf
 * @param {string} caminhoOPF - Caminho completo para o arquivo metadata.opf
 * @returns {Promise<Object|null>} Objeto com dados do livro ou null em caso de erro
 */
async function processarOPF(caminhoOPF) {
    try {
        const xml = fs.readFileSync(caminhoOPF, 'utf8');
        const resultado = await parseStringPromise(xml);
        
        const metadata = resultado.package.metadata[0];
        const dirLivro = path.dirname(caminhoOPF);
        
        const titulo = extrairTitulo(metadata);
        const autor = extrairAutor(metadata);
        const calibreId = extrairCalibreId(metadata);
        
        const capaCaminho = buscarCapa(dirLivro);
        const livroArquivo = buscarArquivoLivro(dirLivro);
        const pastaAutor = path.basename(path.dirname(dirLivro));
        
        return {
            titulo,
            autor,
            calibreId,
            capa: capaCaminho,
            arquivo: livroArquivo,
            pastaAutor,
            caminhoPasta: path.relative(__dirname, dirLivro).replace(/\\/g, '/')
        };
    } catch (erro) {
        console.error(`Erro ao processar ${caminhoOPF}:`, erro.message);
        return null;
    }
}

/**
 * Extrai o título do metadata
 * @param {Object} metadata - Objeto metadata do arquivo OPF
 * @returns {string} Título do livro
 */
function extrairTitulo(metadata) {
    return metadata['dc:title'] ? metadata['dc:title'][0] : 'Título desconhecido';
}

/**
 * Extrai o autor do metadata
 * @param {Object} metadata - Objeto metadata do arquivo OPF
 * @returns {string} Nome do autor
 */
function extrairAutor(metadata) {
    return metadata['dc:creator'] 
        ? metadata['dc:creator'][0]._ || metadata['dc:creator'][0] 
        : 'Autor desconhecido';
}

/**
 * Extrai o ID do Calibre do metadata
 * @param {Object} metadata - Objeto metadata do arquivo OPF
 * @returns {string} ID do Calibre
 */
function extrairCalibreId(metadata) {
    return metadata['dc:identifier']?.find(id => id.$?.scheme === 'calibre')?.$?._ || '';
}

/**
 * Busca o arquivo de capa no diretório do livro
 * @param {string} dirLivro - Diretório do livro
 * @returns {string|null} Caminho relativo da capa ou null
 */
function buscarCapa(dirLivro) {
    const caminhoCompleto = path.join(dirLivro, COVER_FILENAME);
    return fs.existsSync(caminhoCompleto)
        ? normalizarCaminho(path.relative(__dirname, caminhoCompleto))
        : null;
}

/**
 * Busca o arquivo do livro (PDF, EPUB ou MOBI) no diretório
 * @param {string} dirLivro - Diretório do livro
 * @returns {string|null} Caminho relativo do arquivo ou null
 */
function buscarArquivoLivro(dirLivro) {
    const arquivos = fs.readdirSync(dirLivro);
    const arquivo = arquivos.find(f => 
        SUPPORTED_FORMATS.some(ext => f.endsWith(ext))
    );
    return arquivo 
        ? normalizarCaminho(path.relative(__dirname, path.join(dirLivro, arquivo)))
        : null;
}

/**
 * Normaliza o caminho substituindo barras invertidas por barras normais
 * @param {string} caminho - Caminho a ser normalizado
 * @returns {string} Caminho normalizado
 */
function normalizarCaminho(caminho) {
    return caminho.replace(/\\/g, '/');
}

/**
 * Função principal que gera o catálogo de livros
 */
async function gerarCatalogo() {
    console.log('Escaneando pasta Livros...');
    
    const arquivosOPF = buscarArquivosOPF(LIVROS_PATH);
    console.log(`Encontrados ${arquivosOPF.length} livros`);
    
    const livros = [];
    
    for (const opf of arquivosOPF) {
        const dados = await processarOPF(opf);
        if (dados) {
            livros.push(dados);
        }
    }
    
    // Ordenar por título
    livros.sort((a, b) => a.titulo.localeCompare(b.titulo));
    
    // Salvar JSON
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(livros, null, 2), 'utf8');
    
    console.log(`✓ Catálogo gerado com sucesso: ${livros.length} livros`);
    console.log(`✓ Arquivo salvo em: ${OUTPUT_FILE}`);
}

// Executar
gerarCatalogo().catch(console.error);

File: scan-livros.py
#!/usr/bin/env python3
import os
import json
import xml.etree.ElementTree as ET
from pathlib import Path

LIVROS_PATH = Path(__file__).parent / 'Livros'
OUTPUT_FILE = Path(__file__).parent / 'catalogo.json'

def buscar_arquivos_opf(diretorio):
    """Busca recursivamente todos os arquivos metadata.opf"""
    arquivos_opf = []
    
    for root, dirs, files in os.walk(diretorio):
        # Ignorar pasta .caltrash
        if '.caltrash' in root:
            continue
            
        if 'metadata.opf' in files:
            arquivos_opf.append(Path(root) / 'metadata.opf')
    
    return arquivos_opf

def processar_opf(caminho_opf):
    """Extrai informações do arquivo OPF"""
    try:
        tree = ET.parse(caminho_opf)
        root = tree.getroot()
        
        # Namespace do OPF
        ns = {
            'opf': 'http://www.idpf.org/2007/opf',
            'dc': 'http://purl.org/dc/elements/1.1/'
        }
        
        # Extrair metadados
        metadata = root.find('opf:metadata', ns)
        
        titulo = metadata.find('dc:title', ns)
        titulo_texto = titulo.text if titulo is not None else 'Título desconhecido'
        
        autor = metadata.find('dc:creator', ns)
        autor_texto = autor.text if autor is not None else 'Autor desconhecido'
        
        # Buscar descrição
        descricao = metadata.find('dc:description', ns)
        descricao_texto = descricao.text if descricao is not None else ''
        
        # Buscar ID do Calibre
        calibre_id = ''
        for identifier in metadata.findall('dc:identifier', ns):
            if identifier.get('{http://www.idpf.org/2007/opf}scheme') == 'calibre':
                calibre_id = identifier.text
                break
        
        # Diretório do livro
        dir_livro = caminho_opf.parent
        
        # Buscar capa
        caminho_capa = dir_livro / 'cover.jpg'
        capa_relativa = None
        if caminho_capa.exists():
            # Usar caminho relativo ao diretório do script
            capa_relativa = str(caminho_capa.relative_to(LIVROS_PATH.parent)).replace('\\', '/')
        
        # Buscar arquivo do livro (PDF, EPUB, MOBI)
        arquivo_livro = None
        for arquivo in dir_livro.iterdir():
            if arquivo.suffix.lower() in ['.pdf', '.epub', '.mobi']:
                # Usar caminho relativo ao diretório do script
                arquivo_livro = str(arquivo.relative_to(LIVROS_PATH.parent)).replace('\\', '/')
                break
        
        # Pasta do autor (diretório pai)
        pasta_autor = dir_livro.parent.name
        
        return {
            'titulo': titulo_texto,
            'autor': autor_texto,
            'descricao': descricao_texto,
            'calibreId': calibre_id,
            'capa': capa_relativa,
            'arquivo': arquivo_livro,
            'pdf': arquivo_livro,
            'pastaAutor': pasta_autor,
            'caminhoPasta': str(dir_livro.relative_to(Path.cwd())).replace('\\', '/')
        }
        
    except Exception as e:
        print(f"Erro ao processar {caminho_opf}: {e}")
        return None

def gerar_catalogo():
    """Função principal"""
    print("Escaneando pasta Livros...")
    
    arquivos_opf = buscar_arquivos_opf(LIVROS_PATH)
    print(f"Encontrados {len(arquivos_opf)} livros")
    
    livros = []
    
    for i, opf in enumerate(arquivos_opf, 1):
        if i % 50 == 0:
            print(f"Processando livro {i}/{len(arquivos_opf)}...")
        
        dados = processar_opf(opf)
        if dados:
            livros.append(dados)
    
    # Ordenar por título
    livros.sort(key=lambda x: x['titulo'].lower())
    
    # Salvar JSON
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(livros, f, ensure_ascii=False, indent=2)
    
    print(f"\n✓ Catálogo gerado com sucesso: {len(livros)} livros")
    print(f"✓ Arquivo salvo em: {OUTPUT_FILE}")

if __name__ == '__main__':
    gerar_catalogo()
